File: examples/fastvlm/cpp/stream_demo.cpp
/**
 * @file stream_demo.cpp
 * @brief Streaming generation demo for the VLM service.
 *
 * Copyright (C) 2026 SpacemiT (Hangzhou) Technology Co. Ltd.
 * SPDX-License-Identifier: Apache-2.0
 */

#include <cstdio>
#include <iostream>
#include <memory>
#include <string>

#include "vlm/vlm_service.h"

namespace {

std::string GetArg(int argc, char** argv, const std::string& name,
                   const std::string& fallback) {
    for (int i = 1; i + 1 < argc; ++i) {
        if (argv[i] == name) return argv[i + 1];
    }
    return fallback;
}

}  // namespace

int main(int argc, char** argv) {
    const std::string config_path = GetArg(argc, argv, "--config",
        "../examples/fastvlm/config/fastvlm.yaml");
    const std::string image_path = GetArg(argc, argv, "--image", "");
    const std::string prompt = GetArg(argc, argv, "--prompt", "Describe this image");

    std::string error;
    auto service = vlm::CreateVlmServiceFromConfig(config_path, &error);
    if (!service) {
        std::cerr << "Failed to initialize VLM: " << error << std::endl;
        return 1;
    }

    vlm::VlmInput input;
    input.image_path = image_path;
    input.prompt = prompt;

    std::printf("Streaming response:\n");

    vlm::VlmResult result;
    if (!service->GenerateStream(input,
            [](const std::string& chunk, bool is_done, const std::string& err) -> bool {
                if (!err.empty()) {
                    std::fprintf(stderr, "\nStream error: %s\n", err.c_str());
                    return false;
                }
                if (!chunk.empty()) {
                    std::printf("%s", chunk.c_str());
                    std::fflush(stdout);
                }
                return true;
            }, &result, &error)) {
        std::cerr << "\nStream failed: " << error << std::endl;
        return 1;
    }

    std::printf("\n\n--- Stats ---\n");
    std::printf("Latency: %.1f ms\n", result.latency_ms);
    std::printf("Tokens/s: %.1f\n", result.tokens_per_second);
    return 0;
}

File: examples/fastvlm/python/stream_demo.py
#!/usr/bin/env python3
"""Streaming generation demo using the VLM Python bindings.

Demonstrates how to use vlm.VlmService.generate_stream() to receive
tokens incrementally and print them to stdout. Supports SMT/TCM
hardware acceleration.

Usage:
    python3 stream_demo.py --config examples/fastvlm/config/fastvlm.yaml \
        --image /path/to/image.jpg --prompt "Describe this image in detail"

Copyright (C) 2026 SpacemiT
SPDX-License-Identifier: Apache-2.0
"""

import argparse
import sys
import time
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent.parent.parent / "src" / "python"))

from vlm import VlmService


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Streaming generation demo for VLM service"
    )
    parser.add_argument(
        "--config",
        default=str(Path(__file__).parent.parent / "config" / "fastvlm.yaml"),
        help="Path to the YAML configuration file",
    )
    parser.add_argument(
        "--image", default=None, help="Path to an image file (optional)"
    )
    parser.add_argument(
        "--prompt",
        default="Describe this image.",
        help="User prompt text",
    )
    return parser.parse_args()


def main() -> None:
    args = parse_args()

    with VlmService(args.config) as service:
        print("Streaming output:")

        t_start = time.monotonic()
        chunk_count = 0

        for chunk in service.generate_stream(
            args.prompt, image_path=args.image
        ):
            print(chunk, end="", flush=True)
            chunk_count += 1

        elapsed = (time.monotonic() - t_start) * 1000.0

        print(f"\n\n{'='*50}")
        print("  Streaming Statistics")
        print(f"{'='*50}")
        print(f"  Chunks received: {chunk_count}")
        print(f"  Wall time:       {elapsed:.2f} ms")

        metrics = service.get_metrics()
        print(f"  Last latency:    {metrics['last_latency_ms']:.2f} ms")
        print(f"  Last TTFT:       {metrics['last_ttft_ms']:.2f} ms")
        print(f"  Last tokens/sec: {metrics['last_tokens_per_second']:.2f}")


if __name__ == "__main__":
    main()
